### PR TITLE
System SmartContracts + Admin Interface / Implementation

### PR DESCRIPTION
**Pull request by c4t-ag**
### Story:
Camino decided to have some administrative tasks which should be triggered from off-chain services:

1. Camino runs with fixed BaseFee (initially 50GWEI) which should be changeable (e.g. because of the CAM market price).
2. KYC on-chain state will be set from sumsub off chain service, and later distributed to p-chain.
Camino plans to restrict some OPCODEs to the KYC state, for example deploying an solidity smart contract
3. Camino wants to have the option to disable some function calls on existing smart contracts.

Plan is to call all of these admin transactions based on DAO voting results.

Beside this 30% of the transaction fees on c-chain should be collected in a dapp incentive pool which has to be designed later.
Camino already wants a placeholder smart contract for this.

### Implementation:
**The first commit** contains the solidity smart contracts for Admin and IncentivePool, both are coming with a simple access control (role based) and are UpgradeableProxy backed.

4 contracts are predeployed during genesis processing:
- AdminProxy (0x010000000000000000000000000000000000000a)
- AdminImplementation (0x010000000000000000000000000000000000000b)
- IncentiveProxy (0x010000000000000000000000000000000000000c)
- IncentiveImplementation (0x010000000000000000000000000000000000000d)

**The pre-deploy expects an initialAdmin for both contracts which has to be defined in genesis file (cchainConfig).
Without [defining](https://github.com/chain4travel/caminogo/blob/chain4travel/genesis/genesis_columbus.go#L52)"initialAdmin": "0x.....", the default zero address is used and there is no chance to set any roles.**

**The second commit** defines an admin interface in core module, and implements the logic for setting the gasPrice via TX in eth. Idea is to keep as much of Camino specific implementation out of the core.
The current implementation only adds stuff for gasPrice, but now adding new functionality (like KYC) should be possible without major changes.

[The implementation in eth](eth/ethadmin/ethadmin.go) should be quite well cached, for example OPCODE checking should never always read state from disk (if possible).

### Notes:
- Admin settings should be applied on block scope, beginning with the next block which contains the change TX.
- There are no Unit Tests for now which check the new functionality.

For Unit Testing GasPrice following steps has to be done:
- Launch the VM with an initialAdmin where the private key is known
- Give this address (or any other address under control) the GAS_FEE_ROLE by calling grantRole in the Proxy contract, signed by the initialAdmin.
- After this, the adress with GAS_FEE_ROLE rights can call setGasFee (again calling into the proxy!!)

### CODE REVIEW

File: consensus/consensus.go [hunk 1]
```diff
@@ -64,6 +65,9 @@ type ChainHeaderReader interface {
 
 	// GetHeaderByHash retrieves a block header from the database by its hash.
 	GetHeaderByHash(hash common.Hash) *types.Header
+
+	// GetHeaderByHash retrieves a block header from the database by its hash.
```
Comment (DerTiedemann) [inline consensus/consensus.go]: Old Comment

File: consensus/dummy/consensus.go [hunk 2]
```diff
@@ -132,7 +133,8 @@ func (de *DummyEngine) verifyHeaderGasFields(config *params.ChainConfig, header
 	} else {
 		// Verify baseFee and rollupWindow encoding as part of header verification
 		// starting in AP3
-		expectedRollupWindowBytes, expectedBaseFee, err := CalcBaseFee(config, parent, header.Time)
+		// ToDo: AdminController
```
Comment (DerTiedemann) [inline consensus/dummy/consensus.go]: TODO

**Review by peak3d — APPROVED** [review]

File: core/gen_genesis.go [hunk 3]
```diff
@@ -1,14 +1,3 @@
...
-// This file is a derived work, based on ava-labs code whose
-// original notices appear below.
-//
-// It is distributed under the same license conditions as the
-// original code from which it is derived.
-//
-// Much love to the original authors for their work.
-// **********************************************************
-
 // Code generated by github.com/fjl/gencodec. DO NOT EDIT.
```
Comment (DerTiedemann) [inline core/gen_genesis.go:15]: Missing copyright header

File: core/state_processor_test.go [hunk 4]
```diff
@@ -335,7 +335,8 @@ func GenerateBadBlock(parent *types.Block, engine consensus.Engine, txs types.Tr
 		UncleHash: types.EmptyUncleHash,
 	}
 	if config.IsApricotPhase3(new(big.Int).SetUint64(header.Time)) {
-		header.Extra, header.BaseFee, _ = dummy.CalcBaseFee(config, parent.Header(), header.Time)
+		// ToDo: AdminController
+		header.Extra, header.BaseFee, _ = dummy.CalcBaseFee(config, nil, parent.Header(), header.Time)
```
Comment (DerTiedemann) [inline core/state_processor_test.go:339]: TODO

Comment (peak3d) [reply core/state_processor_test.go:339]: @parousisv there is currently no AdminController set up for the unit tests.
While you are working at it, can you pls. implement a dummy one?

**Pull request closed without merging**